File: 02/02-valid-parentheses.js
// https://leetcode.com/problems/valid-parentheses/description/
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = function (s) {
    const pipe = [];
    const correspondingObject = {
        '(': ')',
        '{': '}',
        '[': ']',
    };
    for (let i = 0; i < s.length; i++) {
        const curr = s[i];
        // O(1)
        if (correspondingObject.hasOwnProperty(curr)) {
            // O(n)
            pipe.push(correspondingObject[curr]);
        } else if (pipe.pop() !== curr) {
            return false;
        }
    }
    return pipe.length === 0;
};

// console.log(isValid('(')); //         false
// console.log(isValid('()[]{}')); //    true
// console.log(isValid('(]')); //        false
// console.log(isValid('({([])})')); //  true
// console.log(isValid('({([[])})')); // false
// console.log(isValid(']')); //         false

File: 02-merge-two-sorted-lists/02-valid-parentheses.py
# https://leetcode.com/problems/valid-parentheses/description/

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        correspondingDict = {
            '(': ')',
            '{': '}',
            '[': ']'
        }
        pipe = []
        for c in s:
            if (c in correspondingDict):
                pipe.append(correspondingDict[c])
            elif len(pipe) == 0 or pipe.pop() != c:
                return False
        return len(pipe) == 0

print(Solution().isValid('('))         # False
print(Solution().isValid('()[]{}'))    # True
print(Solution().isValid('(]'))        # False
print(Solution().isValid('({([])})'))  # True
print(Solution().isValid('({([[])})')) # False
print(Solution().isValid(']'))         # False